### ISSUE TITLE
Preview of dot files containing multiple graphs as gif animation

### ISSUE THREAD

**Issue by BoPeng** [vatlab/SoS#919]
Currently our `%preview` magic can preview `.dot` files containing one figure ([source code](https://github.com/vatlab/SoS/blob/master/src/sos/preview.py#L232), [example](https://vatlab.github.io/sos-docs/doc/documentation/SoS_Syntax.html#Mixed-style-workflows--175)).

With the implementation of #888, `.dot` file generated by sos would contain multiple graphs showing the evolution of DAG during the execution of workflows. It would be great if we could preview them as animations (`.gif`), so I believe that we can generate multiple `.png` files using the existing code, and generate `.gif` from these png files.

I look forward to demonstrating SoS using

```
%preview dag.dot
%run -d dag.dot
WORKFLOW
```

Note that displaying gif in Jupyter is possible, but might need some tricks. See Calysto/octave_kernel#105 and ipython/ipython#10045.

@HenryLeongStat , let me know if you are interested in this ticket.

**Comment by HenryLeongStat** [comment]
Interested! Always want to learn how to implement a `magic` 😄

**Comment by BoPeng** [comment]
Great, you can 
1. first have a look at #888 
2. learn how to use the `-d` option to generate `.dot` file containing multiple graphs
3. learn how to generate `.png` files and convert `.png` files to `.gif` using `dot` and then `graphviz/convert`.
4. learn how to convert `dot` to multiple `.png` files programmatically, and convert `.png` to `.gif` using `pygraphviz`.
5. Figure out how to properly preview such `.gif` file inside Jupyter.

Let me know if you need help on any of these.

**Comment by HenryLeongStat** [comment]
Is it basically like the followings? :

  1. Split up a `.dot` file which contains more than 1 `strict digraph{...}` into multiple `.dot` files.
  2. `%preview -n` or `graphviz.Source()` each of them in order to output them as `.png` files. (Is there a way to export the graph as `.png` in `%preview` ? `-s`?)
  3. Combine all the `.png` files as a `.gif` files.
  4. View it. (Tried to use markdown to do it, example: '''
![gundam](https://media.giphy.com/media/eZa8nzRvSln8s/giphy.gif)
''')

**Comment by HenryLeongStat** [comment]
> I have not tried %preview gopher.gif by myself and not sure if %preview can handle `gif files.
Actually, it can! 
<img width="533" alt="screen shot 2018-04-03 at 10 27 43 am" src="https://user-images.githubusercontent.com/26916551/38259040-abcb2a82-3729-11e8-856e-f163f530985b.png">

**Comment by HenryLeongStat** [comment]
Should be very easy! 
Reference: https://stackoverflow.com/questions/753190/programmatically-generate-video-or-animated-gif-in-python

**Comment by HenryLeongStat** [comment]
Also, I tried to figure out whether Source.render can handle multiple chucks of `graph code` ({...}) in a `.dot` file. Seems it is able to do that itself (in CLI), but couldn't find whether it can be done inside Python. I decide to split up the `.dot` files manually, which should be easy I think?

**Comment by HenryLeongStat** [comment]
Done for the splitting:
```
file = open('a.dot', 'r')
text = file.read()
print(text)
splitText = text.split("strict digraph ")
for i in range(1,len(splitText)-1):
	splitName = str(i) + "splitText" + ".dot"
	file = open(splitName, "w")
	writeString = "strict digraph " + splitText[i]
	file.write(writeString)
```

**Comment by BoPeng** [comment]
`Source` can handle strings, no need for temporary files, right?

**Comment by HenryLeongStat** [comment]
Yes, but exporting as `.png` should be in temporary directory. Working on it.

**Comment by HenryLeongStat** [comment]
Opened a new branch and almost finished. 
Professor, how fast you want the `.gif` file changing every scene?

**Comment by HenryLeongStat** [comment]
By default: 
![movie](https://user-images.githubusercontent.com/26916551/38275210-c4b541d6-3756-11e8-907c-85a95ded7e81.gif)

**Comment by BoPeng** [comment]
Looks great. One second for each step?

**Comment by HenryLeongStat** [comment]
One second for each step:
![movie](https://user-images.githubusercontent.com/26916551/38275728-6bf14548-3758-11e8-953f-099243383304.gif)

**Comment by HenryLeongStat** [comment]
In the original code, `preview_dot()` returns as the followings:
```
with open(outfile, 'rb') as content:
       data = content.read()
return {'image/png': base64.b64encode(data).decode('ascii') }
```
What should it return for the `.gif` case?

>Fix the followings:

```
        for filename in pngFiles:
            images.append(imageio.imread(filename))
        data = imageio.mimsave('movie.gif', images)
        os.remove(pngFiles)
        return {'image/gif': movie.gif}
```

**Comment by BoPeng** [comment]
You will need to check current code for previewing images [this line of code](https://github.com/vatlab/SoS/blob/master/src/sos/preview.py#L72) and try. You also need to return regular `png` when there is a single `png` file.

**Comment by HenryLeongStat** [comment]
I see... That's what you are talking about viewing `.gif` in jupyter...

**Comment by HenryLeongStat** [comment]
Super weird... Using the followings code, the static image shows.

```
def preview_dot(filename, kernel=None, style=None):
    from graphviz import Source
    with open(filename) as dot:
        fileNameElement = "sosDotFilesPng"
        src = Source(dot.read(), filename = fileNameElement)
    src.format = 'png'
    outfile = src.render()
    from os import listdir
    from os.path import isfile, join
    import re
    pngFiles = [f for f in listdir("./") if isfile(join("./", f)) and fileNameElement in f and ".png" in f and bool(re.search(r'\d', f))]
    if len(pngFiles == 100):
        with open(outfile, 'rb') as content:
            data = content.read()
        return {'image/png': base64.b64encode(data).decode('ascii') }
    else:
        pngFiles.sort(key=lambda x: int(x.split('.')[1]))
        pngFiles.insert(0, fileNameElement + '.png')
        images = []
        for filename in pngFiles:
            images.append(imageio.imread(filename))
        data = imageio.mimsave('movie.gif', images, duration = 1)
        from wand.image import Image
        img = Image(filename='movie.gif')
        with open('movie.gif', 'rb') as f:
            image = f.read()
        import imghdr
        image_type = imghdr.what(None, image)
        image_data = base64.b64encode(image).decode('ascii')
        os.remove(pngFiles)
        return { 'image/' + image_type: image_data,
                'image/png': base64.b64encode(img._repr_png_()).decode('ascii') }
```

So I remove those two `return` like the followings:
```
def preview_dot(filename, kernel=None, style=None):
    from graphviz import Source
    with open(filename) as dot:
        fileNameElement = "sosDotFilesPng"
        src = Source(dot.read(), filename = fileNameElement)
    src.format = 'png'
    outfile = src.render()
    from os import listdir
    from os.path import isfile, join
    import re
    pngFiles = [f for f in listdir("./") if isfile(join("./", f)) and fileNameElement in f and ".png" in f and bool(re.search(r'\d', f))]
    if len(pngFiles == 100):
        with open(outfile, 'rb') as content:
            data = content.read()
    else:
        pngFiles.sort(key=lambda x: int(x.split('.')[1]))
        pngFiles.insert(0, fileNameElement + '.png')
        images = []
        for filename in pngFiles:
            images.append(imageio.imread(filename))
        data = imageio.mimsave('movie.gif', images, duration = 1)
        from wand.image import Image
        img = Image(filename='movie.gif')
        with open('movie.gif', 'rb') as f:
            image = f.read()
        import imghdr
        image_type = imghdr.what(None, image)
        image_data = base64.b64encode(image).decode('ascii')
        os.remove(pngFiles)
```

It supposed to be not showing up anything, but the image still shows up. How?!?

**Comment by HenryLeongStat** [comment]
Let me use the Linux machine to do the test...

**Comment by HenryLeongStat** [comment]
By the way, professor, it is hard to debug in jupyter notebook... (i.e.: don't know which lines of code were run...)

**Comment by HenryLeongStat** [comment]
Found that there are multiple version of `SoS` installed in my machine:

```
./anaconda/lib/python3.6/site-packages/sos_notebook-0.9.10.12-py3.6.egg/sos_notebook/preview.py:def preview_md(filename, kernel=None, style=None):
./anaconda/lib/python3.6/site-packages/sos-0.9.13.3-py3.6.egg/sos/preview.py:def preview_md(filename, kernel=None, style=None):
./anaconda/lib/python3.6/site-packages/sos-0.9.8.3-py3.6.egg/sos/jupyter/preview.py:def preview_md(filename, kernel=None, style=None):
```

I deleted those old version. Before that, the version run in jupyter is the old version. Not sure why.

**Comment by BoPeng** [comment]
Yes, remove all old versions and try again.

**Comment by HenryLeongStat** [comment]
Btw, finally figured out how to debug in jupyter, and it is fun! :D

**Comment by BoPeng** [comment]
Yes, please use a temporary directory because whatever you use, there is a chance that a user has a valid file in that name...

**Comment by HenryLeongStat** [comment]
OK! Let me fix it.

**Comment by HenryLeongStat** [comment]
Should I create a folder to do it? For now I will get all the files name and then delete them.

**Comment by HenryLeongStat** [comment]
I see! So actually the one created by using `tempfile.TemporaryDirectory()` is different from `tempfile.gettempdir()`. I remember we used `tempfile.gettempdir()` for kernels saving `.feather` files or others.

**Comment by BoPeng** [comment]
Your patch did not work for me. According to https://github.com/ipython/ipython/issues/10045 , the `gif` has to be returned as `image/png` for jupyter to recognize. Can you try the updated patch?

I have also modified your code to use `glob.glob` instead of `os.listdir`, and some other places to make the code shorter. You should learn to use `pyflakes` to find those places though.

**Comment by HenryLeongStat** [comment]
>Your patch did not work for me. According to ipython/ipython#10045 , the gif has to be returned as image/png for jupyter to recognize. Can you try the updated patch?

You mean I put the `.gif` in side image/png as well? And not create any other directory except for **image/png**? Let me update it after the meeing.

>I have also modified your code to use glob.glob instead of os.listdir, and some other places to make the code shorter. You should learn to use pyflakes to find those places though.

OK! Let me find a tutorial of it! 😄

**Comment by HenryLeongStat** [comment]
It works! `duration = 0.5` looks better than 1!

**Comment by BoPeng** [comment]
I have updated the code, just let me know if it works on your end (works for me now).

The problem was that `{'image/gif': gifdata}` does not produce any picture here and I had to use `{'image/png': gifdata}` as suggested by that link.

pyflakes is useful to find places like [here](https://github.com/vatlab/SoS/commit/1f41bbb0e818c1b3fa44a0bb0d78c07a8dc286ea#diff-7173ecb5d3a74747a7d196f1009cc47fL259) when `data` is assigned but never used.
